cranelift/codegen/src/machinst/abi.rs

```diff
@@ -2438,6 +2464,75 @@ impl<M: ABIMachineSpec> CallSite<M> {
     }
 }
 
+impl<T> CallInfo<T> {
+    /// Emit loads for any stack-carried return values using the call
+    /// info and allocations.
+    pub fn emit_retval_loads<
+        M: ABIMachineSpec,
+        EmitFn: FnMut(M::I),
+        IslandFn: Fn(u32) -> Option<M::I>,
+    >(
+        &self,
+        temp: Writable<Reg>,
+        stackslots_size: u32,
+        mut emit: EmitFn,
+        emit_island: IslandFn,
+    ) {
+        // Count stack-ret locations and emit an island to account for
+        // this space usage.
+        let mut space_needed = 0;
+        for CallRetPair { location, .. } in &self.defs {
+            if let RetLocation::Stack(..) = location {
+                // Assume up to ten instructions, semi-arbitrarily:
+                // load from stack, store to spillslot, codegen of
+                // large offsets on RISC ISAs.
+                space_needed += 10 * M::I::worst_case_size();
+            }
+        }
+        if space_needed > 0 {
+            if let Some(island_inst) = emit_island(space_needed) {
+                emit(island_inst);
+            }
+        }
+
+        for CallRetPair { vreg, location } in &self.defs {
+            if let RetLocation::Stack(amode, ty) = location {
+                if let Some(spillslot) = vreg.to_reg().to_spillslot() {
+                    // `temp` is an integer register of machine word
+                    // width, but `ty` may be floating-point/vector,
+                    // which (i) may not be loadable directly into an
+                    // int reg, and (ii) may be wider than a machine
+                    // word. For simplicity, and because there are not
+                    // always easy choices for volatile float/vec regs
+                    // (see e.g. x86-64, where fastcall clobbers only
+                    // xmm0-xmm5, but tail uses xmm0-xmm7 for
+                    // returns), we use the integer temp register in
+                    // steps.
+                    let parts = (ty.bytes() + M::word_bytes() - 1) / M::word_bytes();
+                    for part in 0..parts {
+                        emit(M::gen_load_stack(
+                            amode.offset_by(part * M::word_bytes()),
+                            temp,
+                            M::word_type(),
+                        ));
+                        emit(M::gen_store_stack(
+                            StackAMode::Slot(
+                                i64::from(stackslots_size)
+                                    + i64::from(M::word_bytes())
+                                        * ((spillslot.index() as i64) + (part as i64)),
+                            ),
+                            temp.to_reg(),
+                            M::word_type(),
+                        ));
+                    }
+                } else {
+                    emit(M::gen_load_stack(*amode, *vreg, *ty));
```

Comment:
Could this case assert that `vreg` was not allocated to `temp`?

Reply:
Also, is it possible to somewhat easily assert that `temp` is callee-saved? (or caller-save, I forget which would be appropriate here)

Reply:
Done (to assert), good idea. It's a little trickier to assert anything about whether this is a caller-saved (volatile) register we can use: this helper is called after we've thrown away all the ABI info except for what's in the `CallInfo`, so we'd have to somehow preserve that.